### PR TITLE
Support indexof_n built-in function

### PR DESCRIPTION
**Pull request by shuheiktgw**
Fixes https://github.com/open-policy-agent/opa/issues/4155. This PR adds support for the built-in `indexof_n` function. I'll add the wasm support too once this PR is approved.

### CODE REVIEW

File: topdown/strings.go [hunk 1]
```diff
@@ -87,19 +87,19 @@ func builtinConcat(a, b ast.Value) (ast.Value, error) {
 	return ast.String(strings.Join(strs, string(join))), nil
 }
 
-func builtinIndexOf(a, b ast.Value) (ast.Value, error) {
-	runesEqual := func(a, b []rune) bool {
-		if len(a) != len(b) {
+var runesEqual = func(a, b []rune) bool {
```
Comment (srenatus) [inline topdown/strings.go:90]: This strikes me as a bit uncommon. Why not 
```suggestion
func runesEqual(a, b []rune) bool {
```

**Review by srenatus — COMMENTED** [review]
One oddity in the golang code... Please see inline. That aside, this looks ok.
I'm sure anders would appreciate a Unicode test case 😉

**Comment by anderseknert** [comment]
> One oddity in the golang code... Please see inline. That aside, this looks ok. I'm sure anders would appreciate a Unicode test case 😉

Haha, indeed! It would be good to include at least some simple unicode tests.

File: docs/content/policy-reference.md [hunk 2]
```diff
@@ -371,28 +371,29 @@ complex types.
...
-| <span class="opa-keep-it-together">``output := trim_suffix(string, suffix)``</span> | ``output`` is a ``string`` representing ``string`` with trailing instance of ``suffix`` removed. If ``string`` doesn't end with suffix, ``string`` is returned unchanged.| ✅ |
-| <span class="opa-keep-it-together">``output := trim_space(string)``</span> | ``output`` is a ``string`` representing ``string`` with all leading and trailing white space removed.| ✅ |
-| <span class="opa-keep-it-together">``output := upper(string)``</span> | ``output`` is ``string`` after converting to upper case | ✅ |
+| Built-in                                                                                    | Description                                                                                                                                                                                                                                                                                                            | Wasm Support      |
```
Comment (anderseknert) [inline docs/content/policy-reference.md:374]: I don't think changing the rendering of this entire table (from compact mode to full) is warranted in this PR. Please only add the `indexof_n` row here without changing the rest.

**Review by anderseknert — COMMENTED** [review]
Thanks @shuheiktgw! Looks great. Just a minor remark on the markdown table change, but except for that and a unicode test case or two, this looks good!

File: docs/content/policy-reference.md (hunk 2 above)
Comment (shuheiktgw) [reply docs/content/policy-reference.md]: Sorry, it seems a formatter ran without me noticing it 😅

**Comment by shuheiktgw** [comment]
@srenatus @anderseknert I've updated the PR so would you review it again? 🙏

**Review by anderseknert — APPROVED** [review]
LGTM! Please squash the commits and we should be good to merge this 👍

File: topdown/strings.go [hunk 3]
```diff
@@ -435,6 +467,7 @@ func init() {
 	RegisterFunctionalBuiltin2(ast.FormatInt.Name, builtinFormatInt)
 	RegisterFunctionalBuiltin2(ast.Concat.Name, builtinConcat)
 	RegisterFunctionalBuiltin2(ast.IndexOf.Name, builtinIndexOf)
+	RegisterFunctionalBuiltin2(ast.IndexOfN.Name, builtinIndexOfN)
```
Comment (srenatus) [inline topdown/strings.go:470]: [nit] `RegisterFunctionalBuiltin2` is deprecated, let's use `RegisterBuiltinFunc` instead.

**Review by srenatus — APPROVED** [review]

**Pull request merged**